agda: parameterize agda infrastructure by Agda executable name

carlostome opened the pull request:
This PR addresses two issues:

- `build-support/agda/default.nix` has the name of the executable provided by the `Agda` dependency hardcoded everywhere. This prevents reusing the available infrastructure with other agda executables (e.g., provided by custom agda backends such as  [`agda2hs`](https://github.com/agda/agda2hs)).

- `build-support/agda/default.nix` exports `agda-mode` executable. However, in version 2.8.0 this is deprecated. This PR removes this.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

carlostome committed:
Replace hardcoded agda binary by name from package

carlostome committed:
Remove agda-mode from executables

carlostome commented:
> If not, maybe you can add a derivation to the tests that builds something with agda2hs?

Where would such derivation go?

carlostome commented on pkgs/build-support/agda/default.nix:
```diff
@@ -68,10 +68,9 @@ let
+        makeWrapper ${lib.getExe Agda} $out/bin/${Agda.meta.mainProgram} \
           ${lib.optionalString (ghc != null) ''--add-flags "--with-compiler=${ghc}/bin/ghc"''} \
           --add-flags "--library-file=${libraryFile}"
-        ln -s ${lib.getExe' Agda "agda-mode"} $out/bin/agda-mode
```
The reason is to be compatible with backends that do not have an `agda-mode` executable.

carlostome committed:
Revert change

carlostome committed:
Refactor tests/agda.nix to support more tests

carlostome committed:
Add test to override Agda package with custom backend

philiptaron commented on nixos/tests/all-tests.nix:
```diff
@@ -203,7 +203,9 @@ in
   aesmd = runTestOn [ "x86_64-linux" ] ./aesmd.nix;
   agate = runTest ./web-servers/agate.nix;
-  agda = runTest ./agda.nix;
+  agda = import ./agda {
```
## Tangent about nested attrsets in this file

I see this is following the pattern of the rest of the imports in this file; in other files, I would have expected `lib.recurseIntoAttrs` to make the set of tests here built and eval'd by CI.

I don't really understand @roberth's change https://github.com/NixOS/nixpkgs/pull/191540 that introduced this stanza which deals with that...

https://github.com/NixOS/nixpkgs/blob/f991670eceab6a7dcbb8caea67051b983ed5b031/nixos/tests/all-tests.nix#L113-L125

philiptaron commented on nixos/tests/agda/override-with-backend.nix:
```diff
@@ -0,0 +1,71 @@
+  name = "agda-trivial-backend";
+  meta = {
+    maintainers = [
+      # FIXME
```
Land yourself here?

philiptaron commented on pkgs/build-support/agda/default.nix:
```diff
@@ -45,7 +45,7 @@ let
     let
       libraryFile = mkLibraryFile pkgs;
-      pname = "agdaWithPackages";
+      pname = "${Agda.meta.mainProgram}" + "WithPackages";
```
Is this actually required? It can also be done with one string, rather than concatenation.

```suggestion
      pname = "${Agda.meta.mainProgram}WithPackages";
```

philiptaron reviewed (commented):
I've got a bunch of completely non-blocking comments; my overall comment is that this use of `meta.mainProgram` as the configuration point seems odd and wrong to me.

Isn't the actual contract that whatever replaces `agda` needs to fulfill what it does, even if it uses a different name? 

In other words, if `agda2hs` is used, the trivial derivation that makes it symlink to the `$out/bin/agda` name would _also_ work for your purposes, as I understand it.

carlostome committed:
Update pkgs/build-support/agda/default.nix

Co-authored-by: Philip Taron <philip.taron@gmail.com>

carlostome commented:
> Isn't the actual contract that whatever replaces `agda` needs to fulfill what it does, even if it uses a different name?

To answer this propertly we would have to define first "what (agda) does". For the change I'm proposing I have two cases in mind where one can be argued to fulfill what agda does and the other clearly does not but still can benefit from the nixpkgs infrastructure (`mkDerivation`, `withPackages`):

- Case 1: A backend for Agda that is fully compatible with all Agda libraries present in nixpkgs. 
- Case 2: A backend like `agda2hs` that is not compatible with the Agda libraries in nixpkgs.

> In other words, if `agda2hs` is used, the trivial derivation that makes it symlink to the `$out/bin/agda` name would _also_ work for your purposes, as I understand it.

There is a small subtlety here. Agda backends "know" the name of their executable and use this for example with `agda --emacs-mode setup`. So for instance if we were to use this to drop in a nix shell and use emacs with agda mode then the executable will not be found in the path.

carlostome committed:
Incorporate suggestions

- land `HelloWorld.agda` and `TrivialBackend.hs` in files
- replace `runCommand` by `stdenvNoCC.mkDerivation`
- refactor asserts

carlostome committed:
maintainers: add Carlos Tomé Cortiñas

carlostome committed:
Add myself as a maintainer

carlostome committed:
Refactor agda/base.nix to use HelloWorld.agda file

carlostome committed:
Conditionally create symlink to agda-mode if present
